src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs

```diff
@@ -502,6 +511,9 @@ private void CancellationCallback()
 
                     _wdToPathMap.Clear();
                 }
+
+                // Unlock the root directory
+                Debug.Assert(Interop.Sys.FLock(_inotifyHandle, Interop.Sys.LockOperations.LOCK_UN) == 0);
```

Comment:
Also, though, this doesn't look like the right place to do this cleanup.  What if an exception occurs during processing that causes ProcessEvents to exit but cancellation isn't requested?  Cleanup of the inotify handle is done in the finally block of ProcessEvents... seems like this should be done there, too?